### PR TITLE
PCHR-2148: Allow admin to use leave request popup

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

Admin now can call the Leave Request Popup from Absence tab and use it. The staff contact is now preselected basing on the current contact admin is browsing and cannot be unselected. Admin has the same feature set as manager.

## After

Leave, Sick and TOIL popups are now functioning for admin.

new

## Technical Details

1. New tests added
2. The directive now supports `selected-contact-id` parameter. This parameter is needed to pass the ID of currently browsed staff (contact) by admin. This allows directive to load the needed the needed contact only and preselect it. 
3. request-ctrl.js now has a new property *canManage*, which is true if the role is either manager or admin (both can manage). This property allowed to replace almost all `isRole("manager")` with `canManage`, which allowed to seamlessly provide admin with same functionality and access as manager. toil-request-ctrl.js was also slightly amended, because it also contained `isRole("manager")` checks.
4. Buttons in absence-tab-container.html have been amended to call the popup with right parameters.
5. In the popup views `isRole("manager")` were changed to `canManage`.
6. absence-tab-container.js has to be wrapper to provide `CRM.adminId` (ID of currently logged in admin/user) to the view, which is required by the directive.
7. Red asterisk CSS class was copied in another PR https://github.com/compucorp/civihr-employee-portal-theme/pull/181

## Comments

1. Styles are not consistent between manager's and admin's popups. This is due to the global styling application and not the implementation of the popup.
2. Calling the popup from the admin's calendar is now also supported.
3. Extra bugs found and fixed: 
  - if you selected a From Date not from any absence period (for example, year 1989), then the spinning loader never stopped and JS error happened when you tried to select another date. This bug could also be reproduced in SSP and is fixed now.
  - the Submit button didn't fall into disabled state if you select a non-existing absence period within from/end dates
  - the Submit button didn't recover from disabled state if you select a correct absence period within from/end dates
  - balance was shown even if a non-existing  period was selected
  - if you select any weekend, the day type selector didn't work

---

- [x] Tests Pass

### CODE REVIEW

**Comment by AkA84** [comment]
@igorpavlov 

> A proper "required-mark" classes were applied to mandatory fields ("set status" selector, reason) instead of plain black asterisks.

is this class supported both in the admin and in the SSP?

**Comment by AkA84** [comment]
@igorpavlov next time please make sure to split your work in multiple commits whenever possible, as to provide a nice breakdown of the work the was necessary for your ticket. 11 files in one commit just isn't a good commit history

**Comment by AkA84** [comment]
@igorpavlov 

> The directive now supports `selected-contact-id` param.

What's the purpose of this param? The documentation/explanation is incomplete

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/absence-tab-container.js [hunk 1]
```diff
@@ -1,22 +1,28 @@
...
+/* eslint-env amd */
 
-      $rootScope.role = 'admin';
+(function (CRM) {
+  define([
+    'leave-absences/absence-tab/modules/components'
+  ], function (components) {
+    components.component('absenceTabContainer', {
+      bindings: {
+        contactId: '<'
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/absence-tab-container.js]: are we even using this binding?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/absence-tab-container.js [hunk 2]
```diff
@@ -1,22 +1,28 @@
...
+      controller: ['$log', '$rootScope', function ($log, $rootScope) {
+        $log.debug('Component: absence-tab-container');
 
-      var vm = {};
+        $rootScope.role = 'admin';
 
-      return vm;
-    }]
+        var vm = Object.create(this);
+
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/absence-tab-container.js]: remove empty line

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 3]
```diff
@@ -34,6 +35,7 @@ define([
       this.absencePeriods = [];
       this.absenceTypes = [];
       this.calendar = {};
+      this.canManage = null;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: this is a boolean, so it should be initialized to either `true` or `false`

**Comment by igorpavlov** [comment]
@AkA84 

> is this class supported both in the admin and in the SSP?
Good catch. It doesn't. What should I do in this case?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 4]
```diff
@@ -574,13 +576,20 @@ define([
...
         this.supportedFileTypes = _.keys(sharedSettings.fileUploader.allowedMimeTypes);
         role = this.directiveOptions.userRole || 'staff';
+        this.canManage = this.isRole('manager') || this.isRole('admin');
         this._initRequest();
 
         return loadStatuses.call(self)
           .then(function () {
             initOpenMode.call(self);
 
-            return self.isRole('manager') && loadManagees.call(self);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]:
```js
return this.canManage && loadManagees.call(self);
```
we should keep the number of properties and methods and internal functions to a minimum. Having a `loadManagees()` and a `loadManagee()` functions is confusing and just bloats the component's code. Find an elegant way to handle the exception of having to select only one specific managee in the already existing `loadManagees()` function

**Comment by igorpavlov** [comment]
@AkA84 

> What's the purpose of this param? The documentation/explanation is incomplete

Updated.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 5]
```diff
@@ -965,6 +977,18 @@ define([
       }
 
       /**
+       * Loads just one pre-selected managee
+       *
+       * @return {Promise}
+       */
+      function loadManagee () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: remove this function, see comment above

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/leave-request-ctrl_test.js [hunk 6]
```diff
@@ -387,10 +388,13 @@
             spyOn($ctrl, 'isRole');
           });
 
-          describe('when comment id is missing and role is not manager', function () {
+          describe('when comment id is missing and role is neither manager nor admin', function () {
             beforeEach(function () {
               comment.comment_id = null;
+
               $ctrl.isRole.and.returnValue(false);
+              $ctrl._init();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/leave-request-ctrl_test.js]: weird, why was this added? was the test working at all before?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/absence-tab-container.js (hunk 1 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/absence-tab-container.js]: No we don't. We use `$root.settings.contactId` in templates. I removed it and ran tests - all good.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/absence-tab-container.js (hunk 2 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/absence-tab-container.js]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js (hunk 3 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Done. Also left a comment.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js (hunk 4 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Done. Not sure what "elegancy" could mean here, but I decided to leave it readable.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js (hunk 5 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Done.

**Comment by deb1990** [comment]
@igorpavlov Is the leave modal working as expected when opened by clicking the grid cells of the Calendar page?

**Comment by igorpavlov** [comment]
@deb1990 

> Is the leave modal working as expected when opened by clicking the grid cells of the Calendar page?

checking now

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 7]
```diff
@@ -977,30 +971,25 @@ define([
...
        */
       function loadManagees () {
-        return Contact.find(this.directiveOptions.contactId)
-          .then(function (contact) {
-            return contact.leaveManagees();
-          })
-          .then(function (contacts) {
-            this.managedContacts = contacts;
-          }.bind(this));
+        if (this.directiveOptions.selectedContactId) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Pardon the one liner
```js
function loadManagees () {
  return Contact.find(this.directiveOptions.contactId)
    .then(function (contact) {
      return contact.leaveManagees();
    })
    .then(function (contacts) {
      this.managedContacts = this.directiveOptions.selectedContactId ? _.filter(contacts, function (contact) {
        return // match the contact id with `selectedContactId`
      }) : contacts;
    }.bind(this));
}
```

the logic being

1) get the current contact
2) fetch the managees of the contact
3) if we are specifying a specific managee, then find it in the list of managees
4) otherwise, use the full list

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: @AkA84 Thank you. Could you explain why do we need to fetch full list if we don't need full list in the end? Also, why do we need to make 2 calls instead of one?

**Comment by deb1990** [comment]
civihr_staff_compucorp_co_uk___hr17-1
a: The padding looks broken, it should be like
civihr_staff_compucorp_co_uk___hr17-1

b: its broken

**Comment by deb1990** [comment]
civihr_staff_compucorp_co_uk___hr17-1_
 The margin should not be there

**Comment by deb1990** [comment]
Calendar grid cell still showing leave even after changing the status

1. Create a new Leave Absence
2. Go to Calendar, and open the same and change status to More Information Required and save.
3. See that the calendar reloaded but the Leave still shows but it should not.

This is working fine in SSP manager, could you please investigate this?

**Comment by deb1990** [comment]
civihr_staff_compucorp_co_uk___hr17-1_
The datepicker for expiry date is broken, it can be fixed by using the same markup as From/To date.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 8]
```diff
@@ -574,13 +576,14 @@ define([
 
         this.supportedFileTypes = _.keys(sharedSettings.fileUploader.allowedMimeTypes);
         role = this.directiveOptions.userRole || 'staff';
+        this.canManage = this.isRole('manager') || this.isRole('admin');
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Its better to move 
```
role = this.directiveOptions.userRole || 'staff';
this.canManage = this.isRole('manager') || this.isRole('admin');
```
to a private function with a name like setRoles/initRoles

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/directives/leave-request-popup.js [hunk 9]
```diff
@@ -36,7 +36,8 @@ define([
           contactId: '<',
           leaveRequest: '<',
           leaveType: '@',
-          userRole: '@'
+          userRole: '@',
+          selectedContactId: '<'
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/directives/leave-request-popup.js]: We now have so many variables for contact ID
1. `contactId`
2. `selectedContactId`
3. `adminId`

I am not sure if we need all 3, could you please explain how these 3 are different.

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/absence-tab/components/absence-tab-container.html [hunk 10]
```diff
@@ -1,20 +1,37 @@
-<div class="absence-tab-page">
+<div class="absence-tab-page crm-container">
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/absence-tab/components/absence-tab-container.html]: Why do we need to add `crm-container` class

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js (hunk 7 above)
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: I was about to explain the reasoning behind my solution but then I realized I might have had the wrong assumption. The logic in my solution was that we had to make sure that the given "selected contact" was indeed among the managees of the currently logged in contact, but given that the currently logged in contact in this case is an admin, then the admin is allowed to "use" that contact regardless if he's the contact's l&a manager or not

I think then your solution works in this context

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Thanks.

**Comment by igorpavlov** [comment]
The problem with padding is because of 

```css
#bootstrap-theme .btn {
    border-color: transparent;
    padding: 7px 19px; /* this */
    text-transform: uppercase;
}
```

bootstrap.css?or6lgp:31
#boo

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js (hunk 8 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: @deb1990 could you please explain why would it be better?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/directives/leave-request-popup.js (hunk 9 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/directives/leave-request-popup.js]: `contactId` - currently logged in user
`selectedContactId` - currently being browsed contact (in case of admin)
`adminId` - I need to get in CRM.adminId > absence-tab-container.js > absence-tab-container.html > leave-request-popup call > leave-request-popup.js > request-ctrl.js as `contactId` - changed it to contactId.

However, I share your confusion, I would use:
```js
userId // currently logged in user
contactId // currently being browsed contact
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/absence-tab/components/absence-tab-container.html (hunk 10 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/absence-tab/components/absence-tab-container.html]: I was experimenting with styles and forgot to exclude it, good catch!

**Comment by igorpavlov** [comment]
> The datepicker for expiry date is broken, it can be fixed by using the same markup as From/To date.

Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/directives/leave-request-popup.js (hunk 9 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/directives/leave-request-popup.js]: @igorpavlov Sorry I myself got confused while discussion. I went through your solution and it looks fine to me. No need to change.

**Comment by igorpavlov** [comment]
> a: The padding looks broken, it should have similar padding to

fixed

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/directives/leave-request-popup.js (hunk 9 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/directives/leave-request-popup.js]: @deb1990 no problem, it IS confusing.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js (hunk 8 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Done.

**Comment by igorpavlov** [comment]
> The margin should not be there

fixed

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/absence-tab/components/absence-tab-container.html (hunk 10 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/absence-tab/components/absence-tab-container.html]: Removed

**Comment by igorpavlov** [comment]
Also fixed a bunch of bugs, please see the Comment section.
